compiler/plugins/target/AMD-AIE/aie/AIEDmaToNpu.cpp

```diff
@@ -356,6 +480,39 @@ struct AIEDmaToNpuPass : ::impl::AIEDmaToNpuBase<AIEDmaToNpuPass> {
 
     if (failed(applyPartialConversion(device, target, std::move(patterns))))
       signalPassFailure();
+
+    patterns.clear();
+    target.addIllegalOp<NpuSyncOp>();
+    target.addIllegalOp<NpuWrite32Op>();
+    target.addIllegalOp<NpuAddressPatchOp>();
+    target.addIllegalOp<NpuWriteBdOp>();
+
+    std::vector<uint32_t> instructions;
+    auto words = reserveAndGetTail(instructions, 4);
+
+    // setup txn header
+    words[0] = 0x06030100;
+    words[1] = 0x00000105;
+    uint32_t count = 0;
+
+    patterns.insert<ConvertNpuOp<NpuSyncOp>>(&getContext(), instructions,
+                                             appendSync, count);
+    patterns.insert<ConvertNpuOp<NpuWrite32Op>>(&getContext(), instructions,
+                                                appendWrite32, count);
+    patterns.insert<ConvertNpuOp<NpuAddressPatchOp>>(
+        &getContext(), instructions, appendAddressPatch, count);
+    patterns.insert<ConvertNpuOp<NpuWriteBdOp>>(&getContext(), instructions,
+                                                appendWriteBdShimTile, count);
+
+    if (failed(applyPartialConversion(device, target, std::move(patterns))))
```

Comment:
I think I prefer a non-pattern approach for this. It's not clear here what order the instructions get added (what order the do the patterns get run? I assume it is deterministic, but still). I'm actually surprised this change works with the end-to-end tests, I guess that order in which the pattern manager applies the patterns just happens to be correct for the use case?  

I like the idea of avoiding the llvm::switch that the previous impl had, maybe you can find a solution somewhere between the 2 approaches? We've also had a few discussions in the team with recent PRs about when pattern approaches are appropriate, and when a simple pass if good enough. The general consensus is that if you don't need a fixed-point algorithm (i.e. iterate more than once through the IR until no changes), then we should use a simple pass.

Reply:
The issue is you need `ConversionPattern` in order to be able to have the `eraseOp` actually be applied (or something like that - I always get confused by `ConversionPattern` and it's actually being rewritten [upstream](https://discourse.llvm.org/t/rfc-a-new-one-shot-dialect-conversion-driver/79083)). I had to fiddle to the right thing to happen (the `eraseOp`). 

> I guess that order in which the pattern manager applies the patterns just happens to be correct for the use case?

There's no ordering implied here from the perspective of mlir-aie - it's another one of those spots in AIE dialect where it's just a 1-1 mapping (to register writes). 

**Underlining something here**: there might be an ordering necessitated by the architecture/device itself (e.g., maybe DMAs have to be programmed before locks? or vice-versa?). I have run into issues about ordering stuff (don't enable the cores before starting the DMAs...). **But 100%** whatever order you see in mlir-aie is completely accidental or at best inherited by osmosis from the external examples (some other team) the original code was copy-pasted from. 

Back to mlir itself: in principle the order in which these are applied is determined by the `benefit` (I believe). I don't have a strong opinion on it - I can see why you would prefer something that was more explicitly sequenced but on the other hand there's some (*some*) value in the declarative specification.

> The general consensus is that if you don't need a fixed-point algorithm (i.e. iterate more than once through the IR until no changes), then we should use a simple pass.

Yea I agree with this. Let me try to figure out if I can force the correct `eraseOp` behavior without relying on a `ConversionPattern`.

Reply:
Ok, makes sense now that the order doesn't matter, if it's just register setting (I was confused, thought this was lx6 instructions...)

Sometimes it helps to collect all the ops to erase in a vector while walking the IR, and then remove them as a subsequent step (you've probably seen this design before...). Pointer confusion somewhere, or something.

Reply:
> I was confused, thought this was lx6 instructions

Here is your zen moment: the bulk of LX6 instruction are literally just `write32`, `read32`, `maskwrite32`. So these *are* LX6 instructions.

Reply:
Yeah... makes sense. 

So I'm guessing one of the LX6 instructions is the one which toggles a lock somewhere, setting the AIE in motion, tick tocking away until we've got the result of our kernel. 

And it's important that that instruction appears after all the instructions which initialize BDs, DMAs, etc. 

Is that the point you're making above?

Reply:
Exactly correct - I hit that issue earlier this year with stuff I was doing in mlir-aie+python where I was trying to configure more at runtime. 

This one (program BDs/DMAs before enabling cores) is not such a bad gotcha - it's intuitive. The thing that keeps me up at night is that the DMA/BDs/semaphores can have race conditions. I.e if you sequence your acquires/releases wrong you will get a deadlock. Some of this is documented in the arch spec but some of it is not (ie *bugs in the RTL/hardware..............*).

Reply:
Makes sense. I'm not concerned that there are race conditions and deadlocks, but I haven't actually written any bare metal programs yet...

I still think that we should use a non-pattern approach here for 2 reasons
1) the instruction sequence might change if upstream pattern ordering changes, making the lit test fail which is going to confuse some unfortunate person bumping our llvm in 2 years time
2) there's no fixed point algo here

So let's get back to reconsidering this sometime!

Reply:
I tried to do this this morning before landing. I did a `walk` and erase thing and it failed because there were `uses`. I could've pushed harder ofc but that would mean doing a toposort or slice analysis or something. Which is fine but I felt like I'd just be reproducing some/most of what the `PatternRewriter` infra already does. I also did think about explicitly sequencing the patterns by using `benefit` but then didn't because you have to think a little bit about whether you want to support later inserting a pattern underneath. Anyway combined with the fact that like everything else it'll go away (mostly) I felt like it wasn't worth pursuing. 

So I'll make you a deal: if this code is still here by end of July sure let's redesign but I don't think it will be.

Reply:
Deal.